src/Listeners/Plugin.php

```diff
@@ -44,7 +56,212 @@ public function deactivated( $plugin, $network_wide ) {
+		$plugin['url'] = $data['PluginURI'] ? $data['PluginURI'] : '';
+		return $plugin;
+	}
+
+	/**
+	 * Get plugin dir name or main file if no dir. 
+	 * @param string  $slug Relative path to plugin file
+	 * @return string slug for plugin
+	 */
+	public static function glean_plugin_slugname( $slug ){
+		$newslug = '';
+		// if has `/` split on `/` and get second to last index.
+		if ( str_contains( $slug, '/' ) ) {
+			$parts = explode( '/', $slug );
+			end( $parts );
+			$newslug = prev( $parts );
+		} else {
+			// if not, split on . and get second to last index.
+			$parts = explode( '.', $slug );
+			end( $parts );
+			$newslug = prev( $parts );
+		}
+		return $newslug;
+	}
+
+	/**
+	 * Does this plugin autoupdate?
+	 * @param string  $slug plugin identifier
+	 * @return boolean Whether plugin is configured to auto-update
+	 */
+	public static function does_it_autoupdate( $slug ){
+		// check bluehost plugin setting for auto updates on all plugins
+		$bh_auto_updates = get_site_option( 'auto_update_plugin', 'true' );
+		if ( 'true' === $bh_auto_updates ) {
+			return true;
+		}
+
+		// check core setting for auto updates on this plugin
+		$wp_auto_updates = (array) get_site_option( 'auto_update_plugins', array() );
+		if ( in_array( $slug, $wp_auto_updates, true ) ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Prepare plugin data for single plugin 
+	 * For when single plugin action occurs
+	 * 
+	 * @param string  $slug Name of the plugin
+	 * @param boolean $active Whether the plugin is active
+	 * @param boolean $wrapper Whether to include a plugins wrapper array
+	 * @return Array of data for plugin 
+	 */
+	public static function collect_plugin( $slug, $active, $wrapper=true ){
+		
+		$plugin = [];
+		$pluginpath = WP_PLUGIN_DIR . '/' . $slug;
+
+		if ( ! function_exists( 'get_plugin_data' ) ) {
+			require wp_normalize_path( ABSPATH . '/wp-admin/includes/plugin.php' );
+		}
+		// get data for this plugin
+		$plugin = self::glean_plugin_data( get_plugin_data( $pluginpath ) ); 
+		// key/slug preparations
+		$plugin['slug'] = self::glean_plugin_slugname( $slug );
+		// set other needed data points
+		if ( $active ) {
+			$plugin['active'] = true;
+		}
+		if ( self::does_it_autoupdate( $slug ) ) {
+			$plugin['au'] = true;
+		}
+
+		if ( $wrapper ) {
+			$plugins = [];
+			array_push( $plugins, $plugin );
+			return $plugins;
+		} else {
+			return $plugin;
+		}
+	}
+
+	/**
+	 * Prepare plugin data for all plugins
+	 * @return Array of plugins
+	 */
+	public static function collect_plugins(){
+
+		if ( ! function_exists( 'get_plugins' ) ) {
+			require wp_normalize_path( ABSPATH . '/wp-admin/includes/plugin.php' );
+		}
+		$datas = get_plugins();
+		$mudatas = get_mu_plugins();
+		$plugins = [];
+		
+		// process normal plugins
+		foreach ( $datas as $key => $data ) {
+			$plugin = self::glean_plugin_data( $data ); 
+			// key/slug preparations
+			$plugin['slug'] = self::glean_plugin_slugname( $key );
```

Comment:
Isn't the slug just always going to be `$key`? `get_plugins()` returns an array of plugin data where the keys are the slug.